pkg/syncutil/aggregator/aggregator.go

```diff
@@ -0,0 +1,158 @@
+package aggregator
+
+import (
+	"fmt"
+	gosync "sync"
+
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+// Key defines a type, identifier tuple to store
+// in the GVKAggregator.
+type Key struct {
+	// Source specifies the type or where this Key comes from.
+	Source string
+	// ID specifies the name of the type that this Key is.
+	ID string
+}
+
+func NewGVKAggregator() *GVKAgreggator {
+	return &GVKAgreggator{
+		store:        make(map[Key]map[schema.GroupVersionKind]struct{}),
+		reverseStore: make(map[schema.GroupVersionKind]map[Key]struct{}),
+	}
+}
+
+// GVKAgreggator is an implementation of a bi directional map
+// that stores associations between Key K and GVKs and reverse associations
+// between GVK g and Keys.
+type GVKAgreggator struct {
+	mu gosync.RWMutex
+
+	// store keeps track of associations between a Key type and a set of GVKs.
+	store map[Key]map[schema.GroupVersionKind]struct{}
+	// reverseStore keeps track of associations between a GVK and the set of Key types
+	// that references the GVK in the store map above. It is useful to have reverseStore
+	// in order for IsPresent() and ListGVKs() to run in optimal time.
+	reverseStore map[schema.GroupVersionKind]map[Key]struct{}
+}
+
+// IsPresent returns true if the given gvk is present in the GVKAggregator.
+func (b *GVKAgreggator) IsPresent(gvk schema.GroupVersionKind) bool {
+	b.mu.RLock()
+	defer b.mu.RUnlock()
+
+	_, found := b.reverseStore[gvk]
+	return found
+}
+
+// Remove deletes the any associations that Key k has in the GVKAggregator.
+// For any GVK in the association k --> [GVKs], we also delete any associations
+// between the GVK and the Key k stored in the reverse map.
+func (b *GVKAgreggator) Remove(k Key) error {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	gvks, found := b.store[k]
+	if !found {
+		return nil
+	}
+
+	if err := b.pruneReverseStore(gvks, k); err != nil {
+		return err
+	}
+
+	delete(b.store, k)
+	return nil
+}
+
+// Upsert stores an association between Key k and the list of GVKs
+// and also the reverse associatoin between each GVK passed in and Key k.
+// Any old associations are dropped, unless they are included in the new list of
+// GVKs.
+// It errors out if there is an internal issue with remove the reverse Key links
+// for any GVKs that are being dropped as part of this Upsert call.
+func (b *GVKAgreggator) Upsert(k Key, gvks []schema.GroupVersionKind) error {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	// if any oldOGVKs exist, we deal with 3 flavors of GVKs:
+	// - A. GVKs that are net new
+	// - B. GVKs that are both new and old (intersection)
+	// - C. GVKs that shouold be marked to be pruned because they are no longer
+	// referenced by Key k.
+	//
+	// It is sufficient in this block to determine C and prune GVKs both from
+	// b.store and b.reverseStore. Flavors A & B will be taken care of
+	// in the for loop below.
+	if oldGvks, found := b.store[k]; found {
+		gvksToPrune := b.markOldGVKsToPrune(gvks, oldGvks)
+
+		// prune store
```

Comment:
Perhaps I'm misunderstanding but I believe this is what is required of the logic:

1. Write in the `store` the exact GVKs associated with the K.

2. For the set of GVKs that were removed for K (if any), remove K from each of their value maps in `reverseStore`.

3. Iterate over the GVKs and make sure there is a reference to K for each.

If that's so, I believe the order could be changed to make this a bit simpler:

```golang
// overwrite the GVKs associated with our key
oldGvks := b.store[k]
b.store[k] = gvkSet(gvks)

// remove any reverse references from removed GVKs to this key
gvksRemoved := gvksInFirstButNotSecond(oldGvks, gvks)
b.pruneReverseStore(gvksToPrune, k)

// iterate through they new set of gvks, making sure there is a reference to the key for each
for_, gvk := range gvks {
  if _, found := b.reverseStore[gvk]; !found {
    b.reverseStore[gvk] = make(map[Key]struct{})
  }
  b.reverseStore[gvk][k] = struct{}{}
}
```

Reply:
Thanks for the suggestion! I incorporated some in the re-write